docs/api/vulnerability-info.md

```diff
@@ -0,0 +1,114 @@
+---
+title: Vulnerability Info, NuGet API
+description: The data that allows clients to check packages for known vulnerabilities.
+author: zivkan
+ms.author: zivkan
+ms.date: 06/16/2023
+ms.topic: reference
+---
+
+# Vulnerability information
+
+NuGet client, starting from version 6.7, can download known package vulnerability information to use in scenarios such as checking packages during restore operations.
+
+The API consists of at least two files, the [vulnerability index](#vulnerability-index) and one or more [vulnerability data files](#vulnerability-data).
+Known vulnerability data can be partitioned into multiple files, and the vulnerability index provides clients with information needed to cache files, and update the cache, efficiently.
+
+The resource used for building this URL is the `VulnerabilityInfo` resource found in the [service index](service-index.md).
+
+## Versioning
+
+The following `@type` values are used:
+
+@type value                       | Notes
+--------------------------------- | -----
+VulnerabilityInfo/6.7.0 | The initial release
+
+## Suggested partitioning strategy
+
+The files listed in the vulnerability index should ideally be optimized to maximize caching, and therefore minimize updates to large files.
+This will allow clients to minimize the frequency it needs to download updates.
+
+A suggested strategy for vulnerability data partitioning is to have two files, `base.json` and `updates.json`.
+The `base.json` file is updated periodically (for example once a month), and contains all known vulnerabilities at the time the file is regenerated.
+The `updates.json` file should any new advisories published since `base.json` was last regenerated.
+This will allow clients to download the large `base.json` infrequently, while the frequently changing `updates.json` file is always relatively small.
+
+NuGet client does not provide a way for one file to update or remove entries from another file, so if any modifications are needed to any advisory in `base.json`, then `base.json` can be regenerated and all entries in `updates.json` merged.
+
+## Vulnerability index
+
+The vulnerability index is a JSON array of objects with the following properties:
+
+Name     | Type     | Required | Notes
+-------- | ---------| -------- | -----
+@name    | string   | yes      | A short name for the file, used as a cache key.
+@id      | string   | yes      | The full (absolute) URL to a vulnerability data file.
+@updated | DateTime | yes      | An ISO 8601 string representing the date and time the file was last updated, ideally with the UTC timezone.
+comment  | string   | no       | A optional descriptive string.
+
+The following restrictions apply:
+
+* The index must be an array of objects with between 1 and 16 items.
```

Comment:
Where is this restriction on max 16 enforced. This seems pretty weird to include in the protocol specification since it's not clear to me what goes wrong if this rule is violated (either from the lower bound or upper bound).

Reply:
If there are zero pages, client won't warn or error, but it's bad for customer performance, hence the protocol spec doesn't allow it. If there are more than 16 pages, client will show the customer a warning that their server doesn't implement the protocol correctly, and then will ignore all pages above index 16 (well, 15, using zero indexing).

I don't agree it's weird to specify expected protocol implementations without explaining what the error is. In other documents we don't say "the package ID must be lowercase, and if you don't you're going to get an HTTP 404". Clients can choose to be strict and error out, or can choose to be lenient and allow server implementations that don't follow the spec. But the client does error out, especially if the client starts off lenient and becomes stricter in a future version, at least the protocol was defined precisely from the beginning, so server implementations could have had enough information to be compliant from the beginning.

Reply:
Sure, seems fine!

Reply:
Suppose there is a package source that has a small set of packages released by the feed owners (e.g. a Telerik or some other library vendor). How would they explicitly state to the client that there are no vulnerabilities (yet)? Would it be a single empty page? Or would they opt to not have the resource in the service index at all? The latter option doesn't seem right since that's indistinguishable by the client from the "we don't support that resource" case. It seems to me there may be a scenario for 0 pages if we're talking about a generic protocol, not just what NuGet.org needs (where 0 pages will never be a thing).

Reply:
Does an empty page show more intentionality than a resource with no pages?